### DISCUSSION TITLE
Problem Orm Enum SQLAlchemy v2

### DISCUSSION THREAD

**Discussion by hamedyosefian** [sqlalchemy/sqlalchemy#8856] [Usage Questions]
Hi .. I want create field with type enum but error 
how to do that ?
```
class TypeQuestion(enum.Enum):
    TEST = "Test"
    DESCRIPTIVE = "Descriptive"
```
```
class Question(Base):
    type: Mapped[Enum(TypeQuestion)]
```

**Comment by zzzeek** [comment]
hi -

inside of `Mapped[]`  you put another Python type.  In this case, `TypeQuestion` is a class, which means, it's a type.  You can put it directly inside:

```py
class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    type: Mapped[TypeQuestion]
```

however, SQLAlchemy ~~at the moment does not have an automatic feature~~ (Edit: feature is at  https://docs.sqlalchemy.org/en/20/orm/declarative_tables.html#using-python-enum-or-pep-586-literal-types-in-the-type-map) to link custom enums to the SQLAlchemy ``Enum`` datatype (might be a good idea).  So, you can link TypeQuestion to Enum like this:

```py
#### METHOD ONE - use type_annotation_map ####

class Base(DeclarativeBase):
    type_annotation_map = {
        TypeQuestion: Enum(TypeQuestion)
    }

class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    type: Mapped[TypeQuestion]

```

Or, you can do it like this using Annotated:

```py
#### METHOD TWO - use annotated ####

from typing import Annotated

class TypeQuestion(enum.Enum):
    TEST = "Test"
    DESCRIPTIVE = "Descriptive"

TypeQuestionAnno = Annotated[TypeQuestion, mapped_column(Enum(TypeQuestion))]

class Base(DeclarativeBase):
    pass


class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    type: Mapped[TypeQuestionAnno]

```

**Reply by CaselIT** [reply]
> however, SQLAlchemy at the moment does not have an automatic feature to link custom enums to the SQLAlchemy `Enum` datatype (might be a good idea). 

I guess we can detect subclasses of enum that are not in the type annotation map and automatically map it. If user want to customize the enum they can always specify it explicitly.

I'll open an issue

**Reply by zzzeek** [reply]
yes, though what's the best place to do that, im not sure i want hardcoded rules inside of MappedColumn.

plus people will want to set up options for these types, like native_enum=False.   how to change that across the board?

**Reply by zzzeek** [reply]
it seems like if the Python "Enum" type is part of the type map, we make it so it's that way by default, then have some kind of special lookup if a class is a subclass of Enum, we search for "Enum" directly.  that way it will act like any other type.   obviously looking for the sub-type's entry first before checking Enum.

**Reply by zzzeek** [reply]
ie. walk through `users_enum.__mro__`

**Reply by CaselIT** [reply]
generalizing that is probably not advisable, or we end up mapping datetime to date and similar things we don't want

**Reply by ostapkostyk** [reply]
> > however, SQLAlchemy at the moment does not have an automatic feature to link custom enums to the SQLAlchemy `Enum` datatype (might be a good idea).
> 
> I guess we can detect subclasses of enum that are not in the type annotation map and automatically map it. If user want to customize the enum they can always specify it explicitly.
> 
> I'll open an issue

Are there any news on that? If not, how can I make column type Enum also Optional (nullable=True) if I choose to use Annotated? 
Shall it be like ```TypeQuestionAnno = Annotated[TypeQuestion, mapped_column(Enum(TypeQuestion), nullable=True)]``` and  ```type: Mapped[Optional[TypeQuestionAnno]]``` ?

**Reply by CaselIT** [reply]
it's documented at https://docs.sqlalchemy.org/en/20/orm/declarative_tables.html#using-python-enum-or-pep-586-literal-types-in-the-type-map

**Reply by ostapkostyk** [reply]
I'm sorry, somehow missed this in docs. Still, after reading the docs, I don't understand if Enum can be combined with Optional.
Usually, I use mapped_column(nullable=True) for other types, but in this case, I don't see a way to specify the ```nullable=``` option. And I'm not sure if Optional typing does this.
As I understood, SQLAlchemy checks if the value is in the Enum and raises an exception if it doesn't. So, looks like NULL values would not work, or will they?
Or should I rather make some item in the Enum class like ```NOT_SPECIFIED = None```?

**Reply by zzzeek** [reply]
taking the example, we can modify it to be Optional:

```py
class SomeClass(Base):
    __tablename__ = "some_table"

    id: Mapped[int] = mapped_column(primary_key=True)
    status: Mapped[Optional[Status]]


print(CreateTable(SomeClass.__table__))
```

giving us:

```
CREATE TABLE some_table (
	id INTEGER NOT NULL, 
	status VARCHAR(9), 
	PRIMARY KEY (id)
)

```

or we can use nullable=True with or without optional as:

```py
class SomeClass(Base):
    __tablename__ = "some_table"

    id: Mapped[int] = mapped_column(primary_key=True)
    status: Mapped[Status] = mapped_column(nullable=True)

```

the None/ NULL value is not checked by SQLAlchemy's Enum type as one of the possible values, it passes through as NULL.
